### ISSUE TITLE
Constant errors after setting specific parameters that do not go away unless fully resetting all progress

### ISSUE THREAD

**Issue by JGAAPthrowaway1993** [evllabs/JGAAP#129]
For some background, I am an amateur using JGAAP.

While using JGAAP, I noticed after a day of using it that it started to print out the same error:

`Experiment failed to complete. 
Detailed information is available below: java.lang.Exception: One or more documents could not be read / parsed / canonicized Experiment Failed `

I checked my raw .txt files for the authors to check if there was some sort of corruption preventing them from being read. None were corrupted. 

I wiped all parameters from an analysis I had and inputted fewer parameters, but still nothing.

Currently, the fix is to completely shut off the program, which wipes all of my progress, then adding all authors back and hoping it doesn't come back (and in 1/3 cases it does.)

I noticed with some debugging that the error only occurs when I set specific parameters (Lexical Frequency from the HAL project, Rare Words, most K-steps, and some of the settings related to words at the bottom of the list, and the most perplexing part is that the error still appears even after you remove them from the Event Drivers list??

JGAAP has essentially become unusable for me because I still haven't figured out the rest of the parameters as I use a ton at once and its a gamble on whether I will have to reset my progress.

I don't know where any error logs for this program are, so if anyone can educate me on this first then we can get to troubleshooting. I'm just so lost.

For debugging purposes, I'm on 8.0.2 and on EndeavourOS (a derivative distro of Arch) with the JDK 17 as the main runtime.

Thanks.

**Comment by mvryan** [comment]
Hi

I haven’t worked on this in a very long time but this means for one
document at least it wasn’t able to read it in parse the language of you
have one selected or a canonicizer failed.

Was there any more The the error message ? It should have another error
above this one.
This error is only set here so one of these three logs should be written to
console first that would help identify what’s wrong.
https://github.com/evllabs/JGAAP/blob/96732816bd919292d0b59afff4a9ffbfdb873f02/src/com/jgaap/backend/API.java#L643C21-L643C26

If you’re on Linux these errors should print right to terminal if you
launched with java -jar etc…

Best
Mike

**Comment by JGAAPthrowaway1993** [comment]
Okay, we are getting somewhere and we already close to solving this issue!

The error for every single text file when I selected some parameters was "DefinitionsEventDriver is current not able to run using the jar.  Please use ant with the source distrodution."

I then tested it with every single parameter except the definitions driver to weed out my issues other than the definitions event driver and got these:

```
116893 [pool-2-thread-1] FATAL com.jgaap.backend.API  - Could not load File: /home/XXXXX/JGAAP/XXX (XXX/ZZZZZXX.txt Title:XXX XXXX.txt
java.lang.IllegalArgumentException: Illegal Capacity: -2
        at java.base/java.util.ArrayList.<init>(ArrayList.java:160)
        at com.jgaap.util.EventSet.<init>(EventSet.java:46)
        at com.jgaap.generics.SortedNGramEventDriver.sortEventSet(SortedNGramEventDriver.java:24)
        at com.jgaap.eventDrivers.SortedWordNGramEventDriver.createEventSet(SortedWordNGramEventDriver.java:27)
        at com.jgaap.backend.API$1.call(API.java:635)
        at com.jgaap.backend.API$1.call(API.java:1)
        at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
        at java.base/java.lang.Thread.run(Thread.java:833)
Loading default properties from tagger jar:file:/home/XXX/JGAAP/JGAAP-8.0.2.jar!/com/jgaap/resources/models/postagger/english-left3words-distsim.tagger
Reading POS tagger model from jar:file:/home/XXX/JGAAP/JGAAP-8.0.2.jar!/com/jgaap/resources/models/postagger/english-left3words-distsim.tagger ... done [4.6 sec].
Loading default properties from tagger jar:file:/home/XXX/JGAAP/JGAAP-8.0.2.jar!/com/jgaap/resources/models/postagger/english-left3words-distsim.tagger
Reading POS tagger model from jar:file:/home/XXX/JGAAP/JGAAP-8.0.2.jar!/com/jgaap/resources/models/postagger/english-left3words-distsim.tagger ... M = 1; N = 2


```

```
done [4.4 sec].
M = 1; N = 2
done [1.8 sec].
Loading default properties from tagger jar:file:/home/XXX/JGAAP/JGAAP-8.0.2.jar!/com/jgaap/resources/models/postagger/english-left3words-distsim.tagger
Reading POS tagger model from jar:file:/home/XXX/JGAAP/JGAAP-8.0.2.jar!/com/jgaap/resources/models/postagger/english-left3words-distsim.tagger ... done [3.0 sec].
done [1.7 sec].
Loading default properties from tagger com/jgaap/resources/models/postagger/english-left3words-distsim.tagger
Reading POS tagger model from com/jgaap/resources/models/postagger/english-left3words-distsim.tagger ... M = 1; N = 2
done [1.9 sec].
Loading default properties from tagger jar:file:/home/XXX/JGAAP/JGAAP-8.0.2.jar!/com/jgaap/resources/models/postagger/english-left3words-distsim.tagger
Reading POS tagger model from jar:file:/home/XXX/JGAAP/JGAAP-8.0.2.jar!/com/jgaap/resources/models/postagger/english-left3words-distsim.tagger ... done [2.1 sec].
Loading default properties from tagger com/jgaap/resources/models/postagger/english-left3words-distsim.tagger
Reading POS tagger model from com/jgaap/resources/models/postagger/english-left3words-distsim.tagger ... done [1.7 sec].
M = 1; N = 2
done [1.6 sec].
Loading JAR-internal classifier /com/jgaap/resources/models/ner/english.muc.7class.distsim.crf.ser.gz ... done [1.2 sec].
Loading default properties from tagger jar:file:/home/XXX/JGAAP/JGAAP-8.0.2.jar!/com/jgaap/resources/models/postagger/english-left3words-distsim.tagger
Reading POS tagger model from jar:file:/home/XXX/JGAAP/JGAAP-8.0.2.jar!/com/jgaap/resources/models/postagger/english-left3words-distsim.tagger ... Loading default properties from tagger jar:file:/home/XXX/JGAAP/JGAAP-8.0.2.jar!/com/jgaap/resources/models/postagger/english-left3words-distsim.tagger
Reading POS tagger model from jar:file:/home/XXX/JGAAP/JGAAP-8.0.2.jar!/com/jgaap/resources/models/postagger/english-left3words-distsim.tagger ... done [1.5 sec].
done [1.5 sec].
Loading default properties from tagger jar:file:/home/XXX/JGAAP/JGAAP-8.0.2.jar!/com/jgaap/resources/models/postagger/english-left3words-distsim.tagger
Reading POS tagger model from jar:file:/home/XXX/JGAAP/JGAAP-8.0.2.jar!/com/jgaap/resources/models/postagger/english-left3words-distsim.tagger ... Loading default properties from tagger jar:file:/home/XXX/JGAAP/JGAAP-8.0.2.jar!/com/jgaap/resources/models/postagger/english-left3words-distsim.tagger
Reading POS tagger model from jar:file:/home/XXX/JGAAP/JGAAP-8.0.2.jar!/com/jgaap/resources/models/postagger/english-left3words-distsim.tagger ... done [1.2 sec].
Loading default properties from tagger jar:file:/home/XXX/JGAAP/JGAAP-8.0.2.jar!/com/jgaap/resources/models/postagger/english-left3words-distsim.tagger
Reading POS tagger model from jar:file:/home/XXX/JGAAP/JGAAP-8.0.2.jar!/com/jgaap/resources/models/postagger/english-left3words-distsim.tagger ... done [1.2 sec].
Loading default properties from tagger jar:file:/home/XXX/JGAAP/JGAAP-8.0.2.jar!/com/jgaap/resources/models/postagger/english-left3words-distsim.tagger
Reading POS tagger model from jar:file:/home/XXX/JGAAP/JGAAP-8.0.2.jar!/com/jgaap/resources/models/postagger/english-left3words-distsim.tagger ... done [1.4 sec].
M = 1; N = 2
128021 [pool-2-thread-8] FATAL com.jgaap.backend.API  - Could not load File: /home/XXXXX/JGAAP/Unknown or Alternatives/bannerdomainchange.txt Title:bannerdomainchange.txt
java.lang.IllegalArgumentException: Illegal Capacity: -2
        at java.base/java.util.ArrayList.<init>(ArrayList.java:160)
        at com.jgaap.util.EventSet.<init>(EventSet.java:46)
        at com.jgaap.generics.SortedNGramEventDriver.sortEventSet(SortedNGramEventDriver.java:24)
        at com.jgaap.eventDrivers.SortedWordNGramEventDriver.createEventSet(SortedWordNGramEventDriver.java:27)
        at com.jgaap.backend.API$1.call(API.java:635)
        at com.jgaap.backend.API$1.call(API.java:1)
        at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
        at java.base/java.lang.Thread.run(Thread.java:833)
done [1.2 sec].
M = 1; N = 2
```
(I've censored some info about my personal system with X's and Z's and my project names)

I understand the first error with the definitions, it's just not available, but the second error I can only guess means that there is something in the settings conflicting.

I also noticed that when you press the clear button for the event drivers, it doesn't count it _unless_  you manually click on it to delete it:

Using the "Clear" button:

`227765 [AWT-EventQueue-0] INFO com.jgaap.backend.API  - Adding EventDriver K Skip N Word Gram
229933 [AWT-EventQueue-0] INFO com.jgaap.backend.API  - Adding EventDriver Leave K Out Word NGram
`
Nothing.

Manually deleting by clicking on events:

`314573 [AWT-EventQueue-0] INFO com.jgaap.backend.API  - Removing EventDriver Leave K Out Character NGram
315323 [AWT-EventQueue-0] INFO com.jgaap.backend.API  - Removing EventDriver Leave K Out Character NGram
`

This is some strange behaviour, but for now I'll just avoid all of the borked settings and N-gram settings 'till a resolution is found as a bandaid solution.

**Comment by mvryan** [comment]
Can you take screen shots of how you’re configuring the different steps in
JGAAP?
The current versions are built to take multiple event types at once but if
you’ve loaded everything as one analysis that’s not how it’s intended to be
used.

It looks like something weird is happening in different sorting event
drivers try removing those, and there seems to be a single file of yours it
can’t load correctly

bannerdomainchange.txt

**Comment by JGAAPthrowaway1993** [comment]
I checked and bannerdomainchange.txt and afounder(2).txt were indeed problem files. They were complete duds that I had forgotten to insert the relevant text corpus of (aka, forgot to insert text in the first place). I will close that issue as it is not a failure of the program and is 100% my fault.

I don't enable everything, though I was worried about accurate results. Pre-bug report, this was my stupid setup:

178203 [AWT-EventQueue-0] INFO com.jgaap.backend.API  - Adding Document BannerAlert.txt (unknown)
278027 [AWT-EventQueue-0] INFO com.jgaap.backend.API  - Adding EventDriver Binned Frequencies
279259 [AWT-EventQueue-0] INFO com.jgaap.backend.API  - Adding EventDriver Binned Reaction Times
279917 [AWT-EventQueue-0] INFO com.jgaap.backend.API  - Adding EventDriver Character NGrams
280614 [AWT-EventQueue-0] INFO com.jgaap.backend.API  - Adding EventDriver Characters
282962 [AWT-EventQueue-0] INFO com.jgaap.backend.API  - Adding EventDriver First Word In Sentence
285258 [AWT-EventQueue-0] INFO com.jgaap.backend.API  - Adding EventDriver Lexical Frequencies
288652 [AWT-EventQueue-0] INFO com.jgaap.backend.API  - Adding EventDriver Line Length
290180 [AWT-EventQueue-0] INFO com.jgaap.backend.API  - Adding EventDriver Naming Reaction Times
290721 [AWT-EventQueue-0] INFO com.jgaap.backend.API  - Adding EventDriver New Lines
292441 [AWT-EventQueue-0] INFO com.jgaap.backend.API  - Adding EventDriver Punctuation
294110 [AWT-EventQueue-0] INFO com.jgaap.backend.API  - Adding EventDriver Rare Words
294854 [AWT-EventQueue-0] INFO com.jgaap.backend.API  - Adding EventDriver Sentence Length
296894 [AWT-EventQueue-0] INFO com.jgaap.backend.API  - Adding EventDriver Sentences
298254 [AWT-EventQueue-0] INFO com.jgaap.backend.API  - Adding EventDriver Stanford Part of Speech
298948 [AWT-EventQueue-0] INFO com.jgaap.backend.API  - Adding EventDriver Stanford Part of Speech NGrams
302529 [AWT-EventQueue-0] INFO com.jgaap.backend.API  - Adding EventDriver Word Lengths
303134 [AWT-EventQueue-0] INFO com.jgaap.backend.API  - Adding EventDriver Word NGrams
303654 [AWT-EventQueue-0] INFO com.jgaap.backend.API  - Adding EventDriver Word stems
304212 [AWT-EventQueue-0] INFO com.jgaap.backend.API  - Adding EventDriver Word stems w/ Irregular
305010 [AWT-EventQueue-0] INFO com.jgaap.backend.API  - Adding EventDriver Words
305500 [AWT-EventQueue-0] INFO com.jgaap.backend.API  - Adding EventDriver Words Before and After Named Entities

I had a lot because at the time, I thought that while it was slow, it produced confident results and worked well with the control (random) authors I assigned to see if the scores assigned to the authors I knew must've written the text were higher than controls.

Now because some of the works of a specific author I was searching for have been found and they have vastly different slang terms between the known text and unknown text, I am now using a more minimalist approach that approaches syntax and sentence and word length than searching through n-grams and others. I figured that slang would confuse algorithms for 

The setup for this new approach is: 

923591 [AWT-EventQueue-0] INFO com.jgaap.backend.API  - Adding EventDriver Binned Frequencies
932837 [AWT-EventQueue-0] INFO com.jgaap.backend.API  - Adding EventDriver First Word In Sentence
934855 [AWT-EventQueue-0] INFO com.jgaap.backend.API  - Adding EventDriver Lexical Frequencies
943858 [AWT-EventQueue-0] INFO com.jgaap.backend.API  - Adding EventDriver New Lines
947861 [AWT-EventQueue-0] INFO com.jgaap.backend.API  - Adding EventDriver Punctuation NGrams
949338 [AWT-EventQueue-0] INFO com.jgaap.backend.API  - Adding EventDriver Rare Words
950087 [AWT-EventQueue-0] INFO com.jgaap.backend.API  - Adding EventDriver Sentence Length
950595 [AWT-EventQueue-0] INFO com.jgaap.backend.API  - Adding EventDriver Sentences
954584 [AWT-EventQueue-0] INFO com.jgaap.backend.API  - Adding EventDriver Suffices
955152 [AWT-EventQueue-0] INFO com.jgaap.backend.API  - Adding EventDriver Syllable Transitions
955733 [AWT-EventQueue-0] INFO com.jgaap.backend.API  - Adding EventDriver Syllables Per Word
962250 [AWT-EventQueue-0] INFO com.jgaap.backend.API  - Adding EventDriver Word Lengths


The mode of analysis is usually Markov Chains and Nearest Neighbour, with the distance function usually being TauB, Histo, and correlation distance.

I'm not sure if this what I have done is a good practice or a sin to do with JGAAP, but if it's not a sin then we can consider that closed.

I'm still torn on why the clear button does not let the program know that the event drivers have been wiped by the user. This is the only thing that I am torn on and why I'm not closing the issue for now as it looks pretty misleading or you/the rest of the devs forgot to add it in, which I can't really blame you for :)

I'll look into the source of the clear button error if I can by going into the src if I have time tomorrow.

Thanks for your hard work.

**Comment by JGAAPthrowaway1993** [comment]
Okay, more about the faulty clear button:

At https://github.com/evllabs/JGAAP/blob/96732816bd919292d0b59afff4a9ffbfdb873f02/src/com/jgaap/ui/JGAAP_UI_MainForm.java#L2890C2-L2894C75  (line 2890-2894, couldn't source more than one line for some reason) we have the function regarding the UI element. It tries to call removeAllEventDrivers()

The actual back-end is https://github.com/evllabs/JGAAP/blob/96732816bd919292d0b59afff4a9ffbfdb873f02/src/com/jgaap/backend/API.java#L433C1-L437 (433-437) which governs the actual removing shows that the function between what happens when the user clicks to manually remove and the clear button is totally different.

I noticed that while clicking causes remove.eventdrivers to take into effect, the clearing is managed by something different called removeAllEventDrivers as mentioned before.

I can't investigate since I'm in a rush for unrelated work, but as I said I'll investigate when I am not tired, and I wanted to point you in the right direction without you having to memorize the coding you did years ago.

TL;DR: The program is either calling a function that for some reason is faulty, or it works but a part of the program doesn't seem to understand it.

**Comment by JGAAPthrowaway1993** [comment]
I have marked this as stale because the bug of the clear button is separate from the issue raised here, so I will make a new bug report.
